libslide/src/evaluator_rules/registry/fn_rules.rs

```diff
@@ -14,52 +17,82 @@ macro_rules! get_binary_args {
     };
 }
 
+macro_rules! get_flattened_binary_args {
+    ($expr:expr, $op:expr) => {
+        match $expr.as_ref() {
+            Expr::BinaryExpr(child) if child.op == $op => {
+                Some(get_flattened_binary_args($expr, $op))
+            }
+            _ => None,
+        }
+    };
+}
+
 macro_rules! get_unary_arg {
     ($expr:expr, $op:pat) => {
-        match $expr {
+        match $expr.as_ref() {
             Expr::UnaryExpr(UnaryExpr { op: $op, rhs }) => Some(rhs),
             _ => None,
         }
     };
 }
 
-pub(super) fn add(expr: &Expr) -> Option<Expr> {
-    let (l, r) = get_binary_args!(expr, BinaryOperator::Plus)?;
-    Some(Expr::Const(l + r))
+pub(super) fn add(expr: Rc<Expr>) -> Option<Rc<Expr>> {
+    let mut args = get_flattened_binary_args!(expr, BinaryOperator::Plus)?;
+    let mut konst = 0.;
+    let mut i = 0;
+    for _ in 0..args.len() {
+        match args[i].as_ref() {
+            Expr::Const(f) => {
+                konst += f;
+                args.swap_remove(i);
+            }
+            _ => i += 1,
+        }
+    }
```

Comment:
wait so if you match, we don't increment?

Reply:
Yeah, because if we match then we swap the current index with the last. Then we need to check the current index against to make sure we don't miss anything.

1 x a
^ ---- check 0, add to const

a x
^ ---- we swap removed 0, check a at the same index now